dask/array/rechunk.py

```diff
@@ -569,7 +569,7 @@ def _compute_rechunk(x, chunks):
 
     layer = toolz.merge(x2, intermediates)
     graph = HighLevelGraph.from_collections(merge_name, layer, dependencies=[x])
-    return Array(graph, merge_name, chunks, dtype=x.dtype)
+    return Array(graph, merge_name, chunks, meta=x._meta)
```

Comment:
While we're normalizing things, we might also allow dask arrays to be valid elements.  Maybe something like the following:

```python
def normalize_meta(x, shape=None):
    if isinstance(x, Array):
        return x._meta
    else:
        if shape is None:
            shape = (0,) * x.ndim
        x = x[tuple(...)].reshape(shape)
        return x
```

Then we could pass in `meta=x`.  We do something similar for dask dataframe.

Reply:
And, echoing a point from above, we could call this within the Array constructor to reduce the complexity of the common case

```python
class Array:
    def __init__(self, ..., meta=None):
        meta = normalize_meta(meta)
        ...
```

Reply:
As per my comment in https://github.com/dask/dask/pull/4543#discussion_r283722700, I'll soon check how we can do that normalization in more general terms. Your code suggestion looks really good to me, but I would like to try to handle as many cases as possible in such a `normalize_meta` function, one other case is for example to reduce it to `0`d, which I've been doing by calling `np.sum` upon.